chainer/optimizer.py

```diff
@@ -712,6 +712,15 @@ def update_loss_scale(self):
         self._loss_scale = max(1, min(self._loss_scale_max,
                                       self._loss_scale * multiplier))
 
+    def multidevice_hyperparameter_correction(self, num_devices):
```

Comment:
I suspect the name `multidevice_hyperparameter_correction` is not something we really want to represent.

What we want to do is "to adapt the hyperparameters for modified batch size".
It's better to come up with a name that express that.

`adjust_hyperparameters_for_modified_batch_size`?
Or perhaps simply `adjust_hyperparameters`? But I think explicitness is more important here.

Also, the argument should be a real number value, like `factor` (= `1 / num_devices`).

Reply:
Or how about introducing something like `Optimizer.batch_size_factor` and `UpdateRule.batch_size_factor` and referring to the latter from `UpdateRule.update_core()`?
I think it must be implemented as a hyperparameter in order for it to be correctly propagated from `Optimizer` to `UpdateRule`.

Reply:
You mean, doing the hyperparameter adjust inside `UpdateRule.update_core()`?

Reply:
Yes. Not rewriting the hyperparameter indefinitely, but just using an adjusted hyperparameter for each single update.
This will be no-compat because the hyperparameter values exposed to users will change (currently adjusted values are exposed, but after this change, they will be non-adjusted values).

Reply:
One of the things that I see in this case, is that we need to write explicit UpdateRules for all the optimizers. In the case of Adam, the derived optimizers share the `AdamUpdateRule`, but they adjust the hyper-parameters differently.

I am not against this change as a think the potential benefits compensate the extra verbosity.

Reply:
By "vebosity" do you mean the necessity to implement separate `UpdateRule`s?
I don't think this is necessary. `AdamRule` can implement different adjustment formulas for different modes (amgrad / adabound etc.) in a single class.
Even if we need separate `UpdateRule` classes, it's not necessarily a bad thing. The code will be more organized.

Reply:
Yes, I tried to update the comment and blew it. Today is not my day

Reply:
If we go with the hyperparameter approach it seems that we need to pass`batch_size_factor` as a parameter of the optimizer constructor.
This will make the user to explicitly define it when instantiating the optimizer. Do you have any idea about how to solve this?

Reply:
Hmm, I don't understand what the problem is.
As you said it can be assigned via a property `optimizer.batch_size_factor`, not via constructor.
But what does this part mean?

> then override `update_core` for the `UpdateRule` to set the actual value.

What's "the actual value" referring to? `optimizer.batch_size_factor`?

Reply:
the learning rate or eps value adjusted with the batch_size_factor

I feel like this can be done with a regular property to set the eps or the lr instead of a hyperparameter.

Reply:
Do you mean reassigning calculated hyperparameters as attributes of `UpdateRule`?
I don't think this is needed. We can calculate the adjusted hyperparameters every time `update_core` is called.

Reply:
I just added a commit that does it for optimizer.py and ada_grad.py
It is an example and not complete, but I would like to know if this is the right way to do it.
Thanks!

Reply:
I left some comments about the implementation.

Reply:
I'm just beginning to feel `batch_size_factor` might be too specific too.
How about `grad_scale_factor`, `loss_scale_factor` (perhaps confusing?), etc.?

Reply:
I think that `grad_scale_factor` or `loss_scale_factor` gives the wrong idea  that the parameter is used to alter the grads or the loss value somehow. I think that `updater_scale_factor` may be generic enough.

Reply:
Updaters are just an instance of users of optimizers. I think it would be a reverse dependency to name it based on updaters.

Reply:
We have discussed offline and found that there are no better fitting alternatives than `grad_scale_{factor, compensator}` for this use case.

Reply:
hyperparameter renamed